MSD - Stats Card - Handle Success Response

ashiagr opened the pull request:
Parent https://github.com/wordpress-mobile/WordPress-Android/issues/15749

This is a draft PR that includes the following changes for adding the `Stats Card` to the `cards` result:
- JSON deserialization for the success response 
- Mapping success response to model

----------------------------------------

There's nothing to test in this PR as the `dashboard/cards` endpoint doesn't yet include the `Stats Card` yet. So that existing tests work as expected,  I've updated the following test classes: 
- `CardsRestClientTest` is updated to add `StatsResponse`and `cards.json` is updated to include the `Stats Card`.
- `CardsStoreTest` is updated to add `STATS_ENTITY` to `CARDS_ENTITY`.
----------------------------------------

Note: We might want to keep this PR in draft mode until the `dashboard/cards` endpoint is updated for the `Stats Card` as the final structure might vary. With the current prod version of the endpoint, the stats card response being `null` is not added to the cards list.

ashiagr committed:
Feature: Create stats card response and model

ashiagr committed:
Tests: Update cards rest client tests to include stats card response

ashiagr committed:
Tests: Update cards store tests to include stats card entity

zwarm commented:
> We might want to keep this PR in draft mode

@ashiagr , yes, let's do that! 
(1) I added `comments` to the endpoint because it's needed for iOS; not sure if we want to do anything with it. 
(2) Let's sync on what the success looks like ... I was basing it off of the following, but that can be changed.
```

Success: return array(
        'ok'           => true,
	'error'      => '',
	'stats'      => stats data,
      );
Failure: return array(
        'ok'     => false,
	 'error' => 'error message',
      );
```
NOTE:: We decided against the above approach, this is here for history purposes

ashiagr committed:
Merge branch 'trunk' into issue/15749-msd-stats-card-handle-success-response

ashiagr committed:
Refactor: Update dashboard/cards endpoint success response.
- Added comments to todays_stats.
- Moved nested todays_stats to keep it at the same level as posts node in cards.json and corresponding response and model classes.

Note that the existing endpoint will not return the stats card yet. It will be switched to a newer version in future commits.

zwarm commented:
@ashiagr - The `cards-data` endpoint has been deployed - you can swap out the mock data responses.

ashiagr committed:
Replace dashboard cards endpoint

ashiagr committed:
Pass card types to fetch cards

ashiagr committed:
Pass cardTypes to getCards

ashiagr committed:
Tests: Add tests for fetchCards with different card types

ashiagr committed:
Merge branch 'trunk' into issue/15749-msd-stats-card-handle-success-response

zwarm reviewed (approved):
@ashiagr Very well done 👏 

❤️  for the patience in changing all those tests.

I like the improvement that added "label" to types so the arguments build out works smoothly.

I tested using the WPAndroid#[15893 ]([MSD - Handle Single Dashboard Card Error in UI](https://github.com/wordpress-mobile/WordPress-Android/pull/15893#)) and was able to view the stats card as expected. I also visited a few web sites and PTR's, and the counts updated. 

Please feel free to merge when you feel comfortable.
